core/src/main/java/cz/o2/proxima/scheme/SchemaDescriptors.java

```diff
@@ -0,0 +1,885 @@
+      if (!super.equals(o)) {
+        return false;
+      }
+      @SuppressWarnings("unchecked")
+      StructureTypeDescriptor<?> that = (StructureTypeDescriptor<?>) o;
+      return name.equals(that.name) && fields.equals(that.fields);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(super.hashCode(), name, fields);
+    }
+  }
+
+  /**
+   * Enum type descriptor.
+   *
+   * @param <T> value type
+   */
+  public static class EnumTypeDescriptor<T extends Serializable> extends GenericTypeDescriptor<T>
+      implements WithDefaultValue<T> {
+
+    @Getter private final List<T> values;
+    @Nullable private final Factory<T> defaultValueFactory;
+
+    public EnumTypeDescriptor(List<T> values, @Nullable Factory<T> defaultValueFactory) {
+      super(AttributeValueType.ENUM);
+      this.values = values;
+      this.defaultValueFactory = defaultValueFactory;
+    }
+
+    @Nullable
+    @Override
+    public T getDefaultValue() {
+      return defaultValueFactory != null ? defaultValueFactory.apply() : null;
+    }
+
+    @Override
+    public SchemaTypeDescriptor<T> toTypeDescriptor() {
+      return SchemaTypeDescriptor.<T>builder().type(getType()).enumTypeDescriptor(this).build();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      if (!super.equals(o)) {
+        return false;
+      }
+      @SuppressWarnings("unchecked")
+      EnumTypeDescriptor<?> that = (EnumTypeDescriptor<?>) o;
+      return values.equals(that.values);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(super.hashCode(), values);
+    }
+  }
+
+  @ToString
+  public static class FieldTypeDescriptor<T> extends GenericTypeDescriptor<T>
+      implements WithDefaultValue<T> {
+
+    @Getter final String name;
+    @Getter final SchemaTypeDescriptor<T> valueDescriptor;
+    @Nullable final Factory<T> defaultValueFactory;
+
+    public FieldTypeDescriptor(
+        String name, SchemaTypeDescriptor<T> descriptor, @Nullable Factory<T> defaultValueFactory) {
+      super(descriptor.getType());
+      this.name = name;
+      this.valueDescriptor = descriptor;
+      this.defaultValueFactory = defaultValueFactory;
+    }
+
+    public AttributeValueType getValueType() {
+      if (valueDescriptor.getType().equals(AttributeValueType.ARRAY)) {
+        // for arrays we want to return inner type
+        return valueDescriptor.getArrayTypeDescriptor().getValueType();
+      }
+      return valueDescriptor.getType();
+    }
+
+    @SuppressWarnings("squid:S1452")
+    public FieldTypeDescriptor<?> getField(String name) {
+      Preconditions.checkState(
+          getValueType().equals(AttributeValueType.STRUCTURE),
+          "Field type is not a structure. Type: " + getValueType());
+      return valueDescriptor.getStructureTypeDescriptor().getField(name);
```

Comment:
Uff ... `return this`? What is the purpose of this method?

Reply:
intention of this was allow to call something like:
`structure.getField("userInfo").getField("id")`
basically get another field from field which is STRUCTURE but it can be done better, specially if FieldTypeDescriptor will be removed.

Reply:
Method removed as it doesn't make sense